File: learning-factory-integration/warehouse/artifacts/projects/languages/sprig-bytecode-vm/job_project_bytecode_vertical_v1/attempt-002/alternatives/treewalk/tinyvm/interpreter.py
from __future__ import annotations

from .model import Assign, Binary, Block, ExecutionResult, If, Let, Literal, Print, Program, ResourceLimit, RuntimeFault, Unary, Variable, While
from .semantics import binary, checked, truth


class Interpreter:
    def __init__(self, max_steps: int):
        if isinstance(max_steps, bool) or not isinstance(max_steps, int) or max_steps <= 0:
            raise ValueError("max_steps must be a positive integer")
        self.max_steps = max_steps
        self.steps = 0
        self.environment: dict[str, int] = {}
        self.outputs: list[int] = []

    def tick(self) -> None:
        if self.steps >= self.max_steps:
            raise ResourceLimit(f"evaluation budget exceeded ({self.max_steps})")
        self.steps += 1

    def run(self, program: Program) -> ExecutionResult:
        for statement in program.statements:
            self.statement(statement)
        return ExecutionResult(tuple(self.outputs), dict(self.environment), self.steps, "treewalk")

    def statement(self, statement) -> None:
        self.tick()
        if isinstance(statement, Let):
            value = self.expression(statement.initializer)
            if statement.name in self.environment:
                raise RuntimeFault(f"duplicate variable: {statement.name}")
            self.environment[statement.name] = value
        elif isinstance(statement, Assign):
            value = self.expression(statement.value)
            if statement.name not in self.environment:
                raise RuntimeFault(f"undefined variable: {statement.name}")
            self.environment[statement.name] = value
        elif isinstance(statement, Print):
            self.outputs.append(self.expression(statement.expression))
        elif isinstance(statement, Block):
            for child in statement.statements: self.statement(child)
        elif isinstance(statement, If):
            branch = statement.then_branch if self.expression(statement.condition) else statement.else_branch
            if branch is not None: self.statement(branch)
        elif isinstance(statement, While):
            while self.expression(statement.condition): self.statement(statement.body)
        else:
            raise RuntimeFault(f"unsupported statement: {type(statement).__name__}")

    def expression(self, expression) -> int:
        self.tick()
        if isinstance(expression, Literal): return checked(expression.value)
        if isinstance(expression, Variable):
            if expression.name not in self.environment:
                raise RuntimeFault(f"undefined variable: {expression.name}")
            return self.environment[expression.name]
        if isinstance(expression, Unary):
            if expression.operator == "-" and isinstance(expression.right, Literal) and expression.right.value == 2 ** 63:
                self.tick()
                return -(2 ** 63)
            value = self.expression(expression.right)
            if expression.operator == "-": return checked(-value)
            if expression.operator == "!": return int(value == 0)
        if isinstance(expression, Binary):
            left = self.expression(expression.left)
            if expression.operator == "&&": return truth(self.expression(expression.right)) if left else 0
            if expression.operator == "||": return 1 if left else truth(self.expression(expression.right))
            return binary(expression.operator, left, self.expression(expression.right))
        raise RuntimeFault(f"unsupported expression: {type(expression).__name__}")


def execute(program: Program, *, max_steps: int = 10_000) -> ExecutionResult:
    return Interpreter(max_steps).run(program)

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/27debb2830b4885fa831/repair-v1/job_byox_repair_v1_g1_f7de7abfdff812ca5a1921818f925788/attempt-001/sealed/reference/src/interpreter.js
import { RuntimeError, boundedInteger } from "./errors.js";
import { binary, formatValue, isLanguageValue, isTruthy, unary } from "./semantics.js";

const MAX_STEPS = 100_000;

export function interpret(ast, options = {}) {
  const maxSteps = boundedInteger(options, "maxSteps", MAX_STEPS);
  return new Interpreter(maxSteps).run(ast);
}

class Environment {
  constructor(parent = null) {
    this.parent = parent;
    this.values = new Map();
  }

  define(name, value, loc) {
    if (this.values.has(name)) throw new RuntimeError(`Variable '${name}' is already defined`, loc);
    this.values.set(name, value);
    return value;
  }

  get(name, loc) {
    if (this.values.has(name)) return this.values.get(name);
    if (this.parent) return this.parent.get(name, loc);
    throw new RuntimeError(`Undefined variable '${name}'`, loc);
  }

  assign(name, value, loc) {
    if (this.values.has(name)) {
      this.values.set(name, value);
      return value;
    }
    if (this.parent) return this.parent.assign(name, value, loc);
    throw new RuntimeError(`Undefined variable '${name}'`, loc);
  }
}

class Interpreter {
  constructor(maxSteps) {
    this.maxSteps = maxSteps;
    this.steps = 0;
    this.output = [];
  }

  run(ast) {
    if (!ast || ast.type !== "Program" || !Array.isArray(ast.body)) {
      throw new RuntimeError("Expected a Program AST", ast?.loc);
    }
    const global = new Environment();
    let value = null;
    for (const statement of ast.body) value = this._execute(statement, global);
    return { value, output: [...this.output] };
  }

  _execute(node, environment) {
    this._tick(node?.loc);
    switch (node?.type) {
      case "LetStatement": {
        const name = requireIdentifier(node.name, node.loc);
        const value = this._evaluate(node.initializer, environment);
        return environment.define(name, value, node.name.loc);
      }
      case "PrintStatement": {
        const value = this._evaluate(node.expression, environment);
        this.output.push(formatValue(value));
        return value;
      }
      case "ExpressionStatement":
        return this._evaluate(node.expression, environment);
      case "BlockStatement":
        return this._block(node, new Environment(environment));
      case "IfStatement":
        if (isTruthy(this._evaluate(node.test, environment))) {
          return this._execute(node.consequent, environment);
        }
        return node.alternate ? this._execute(node.alternate, environment) : null;
      case "WhileStatement": {
        let value = null;
        while (isTruthy(this._evaluate(node.test, environment))) {
          value = this._execute(node.body, environment);
        }
        return value;
      }
      default:
        throw new RuntimeError(`Unknown statement node ${JSON.stringify(node?.type)}`, node?.loc);
    }
  }

  _block(node, environment) {
    if (!Array.isArray(node.body)) throw new RuntimeError("Block body must be an array", node.loc);
    let value = null;
    for (const statement of node.body) value = this._execute(statement, environment);
    return value;
  }

  _evaluate(node, environment) {
    this._tick(node?.loc);
    switch (node?.type) {
      case "Literal":
        if (!isLanguageValue(node.value)) throw new RuntimeError("Invalid literal value", node.loc);
        return node.value;
      case "Identifier":
        return environment.get(requireIdentifier(node, node.loc), node.loc);
      case "AssignmentExpression": {
        const value = this._evaluate(node.value, environment);
        return environment.assign(requireIdentifier(node.name, node.loc), value, node.name.loc);
      }
      case "UnaryExpression":
        return unary(node.operator, this._evaluate(node.argument, environment), node.loc);
      case "BinaryExpression": {
        return this._binaryChain(node, environment);
      }
      case "LogicalExpression": {
        return this._logicalChain(node, environment);
      }
      default:
        throw new RuntimeError(`Unknown expression node ${JSON.stringify(node?.type)}`, node?.loc);
    }
  }

  _binaryChain(root, environment) {
    const chain = [];
    let expression = root;
    let first = true;
    while (expression && expression.type === "BinaryExpression") {
      if (first) first = false;
      else this._tick(expression.loc);
      chain.push(expression);
      expression = expression.left;
    }
    let value = this._evaluate(expression, environment);
    for (let index = chain.length - 1; index >= 0; index -= 1) {
      const node = chain[index];
      value = binary(node.operator, value, this._evaluate(node.right, environment), node.loc);
    }
    return value;
  }

  _logicalChain(root, environment) {
    const chain = [];
    let expression = root;
    let first = true;
    while (expression && expression.type === "LogicalExpression") {
      if (first) first = false;
      else this._tick(expression.loc);
      chain.push(expression);
      expression = expression.left;
    }
    let value = this._evaluate(expression, environment);
    for (let index = chain.length - 1; index >= 0; index -= 1) {
      const node = chain[index];
      if (node.operator === "or") {
        if (!isTruthy(value)) value = this._evaluate(node.right, environment);
      } else if (node.operator === "and") {
        if (isTruthy(value)) value = this._evaluate(node.right, environment);
      } else {
        throw new RuntimeError(`Unknown logical operator ${JSON.stringify(node.operator)}`, node.loc);
      }
    }
    return value;
  }

  _tick(loc) {
    this.steps += 1;
    if (this.steps > this.maxSteps) {
      throw new RuntimeError(`Execution step limit ${this.maxSteps} exceeded`, loc);
    }
  }
}

function requireIdentifier(node, loc) {
  if (!node || node.type !== "Identifier" || typeof node.name !== "string" ||
      !/^[A-Za-z_][A-Za-z0-9_]*$/.test(node.name)) {
    throw new RuntimeError("Expected an identifier", loc);
  }
  return node.name;
}
